mgosimplekv/kv.go

```diff
@@ -0,0 +1,171 @@
+package mgosimplekv
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	errgo "gopkg.in/errgo.v1"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+	retry "gopkg.in/retry.v1"
+
+	"github.com/juju/simplekv"
+)
+
+// kvStore implements simplekv.Store.
+type kvStore struct {
+	coll *mgo.Collection
+}
+
+// NewStore returns a new Store implementation that uses
+// the given mongo collection for storage.
+func NewStore(coll *mgo.Collection) (simplekv.Store, error) {
+	if err := coll.EnsureIndex(mgo.Index{
+		Key:         []string{"expire"},
+		ExpireAfter: time.Second,
+	}); err != nil {
+		return nil, errgo.Mask(err)
+	}
+	return &kvStore{
+		coll: coll,
+	}, nil
+}
+
+type sessionKey struct{}
+
+// Context implements simplekv.Context by copying the kvStore's underlying
+// session if one isn't already present in the context.
+func (s *kvStore) Context(ctx context.Context) (_ context.Context, close func()) {
+	if session, _ := ctx.Value(sessionKey{}).(*mgo.Session); session != nil {
+		return ctx, func() {}
+	}
+	// TODO add some more advanced session pooling.
```

Comment:
I'd remove this TODO. I don't think session pooling is this library's place. Better would be a public method that allowed a session to be added to a context so that an application can manage sessions externally.

Reply:
I changed the TODO.